### PR TITLE
feat(ertp): support min/max for AmountMath

### LINKED ISSUE #4397: Consider adding amountMath.min() and amountMath.max()
As suggested by @JimLarson at

https://github.com/Agoric/agoric-sdk/pull/4361#issuecomment-1020616357

### PR DESCRIPTION
**Pull request by b4d2**
closes: #4397 

## Description

Adds min/max APIs to ERTP AmountMath 

### Security Considerations

The behavior of the set math min/max (count) may or may not be useful, and can be migrated to NaN if that makes more sense.

### Documentation Considerations

This PR would need a followup PR to add the new APIs to docs to fully close the issue once the API is finalized

### Testing Considerations

I've added unit tests alongside the existing AmountMath unit tests

### CODE REVIEW

**Review by erights — CHANGES_REQUESTED** [review]
Just putting a "Request changes" on this to ensure it doesn't go forward in its current state, until I can engage. Basic objection is that the inequality it should be based on is amountMath.isGTE, which is already abstracted across kinds, rather than do something like comparing lengths.

Please add both @JimLarson and I as reviewers.

**Comment by b4d2** [comment]
That makes sense - I've reworked the SET-related implementations to effectively alias the existing superset logic used in isGTE.  For assignees, I don't have perms but I can ask around.

**Comment by dckc** [comment]
> ... the inequality it should be based on is amountMath.isGTE, which is already abstracted across kinds, rather than do something like comparing lengths.

I made a rough suggestion to that effect: https://github.com/b4d2/agoric-sdk/pull/1

File: packages/ERTP/test/unitTests/mathHelpers/test-copyBagMathHelpers.js [hunk 1]
```diff
@@ -170,6 +170,29 @@ test('copyBag with strings add', t => {
   );
 });
 
+test('copyBag with strings min/max', t => {
```
Comment (dckc) [inline packages/ERTP/test/unitTests/mathHelpers/test-copyBagMathHelpers.js]: I wonder about the value of min/max tests in each mathHelpers; they're redundant with `isGTE` tests (or should be).

A test to verify the availability of the `min/max` API should suffice, I think.

Comment (dckc) [reply packages/ERTP/test/unitTests/mathHelpers/test-copyBagMathHelpers.js]: One thing that would be quite nice is to take the property-based testing for bag amounts #4366 and generalize to all amounts.

Comment (erights) [reply packages/ERTP/test/unitTests/mathHelpers/test-copyBagMathHelpers.js]: > I wonder about the value of min/max tests in each mathHelpers; they're redundant with `isGTE` tests (or should be).
> 
> A test to verify the availability of the `min/max` API should suffice, I think.

I don't understand either of these comments. The way the PR looks now, with `min` and `max` defined once, simply, in terms of `isGTE` seems exactly right. We can leave the helpers out of this.

Comment (dckc) [reply packages/ERTP/test/unitTests/mathHelpers/test-copyBagMathHelpers.js]: What I mean is: it's worth having a test that verifies that `AmountMath.min` exists and works as designed in a simple case or two; likewise `AmountMath.max`. That's what I meant by a test to verify the availability of the min/max API.

Lots of tricky min/max tests don't seem worthwhile; the definition of min/max in terms of isGTE is sufficiently simple that any effort one might put into testing min/max should be put into testing isGTE instead.

**Review by erights — APPROVED** [review]
All the non-test changes LGTM. I did not review the test changes in detail but didn't see anything alarming. I suggest you and @dckc converge on review of tests though, before merging.

@dckc I love the idea of making the property-based testing general across all of AmountMath. But does it need to be in this PR? It seems like a distinct piece of work.

**Comment by dckc** [comment]
> @dckc I love the idea of making the property-based testing general across all of AmountMath. But does it need to be in this PR?

no.

> It seems like a distinct piece of work.

yes.

**Review by dckc — CHANGES_REQUESTED** [review]
please prune the MathHelpers level tests

**Comment by b4d2** [comment]
Pruned the NAT tests

**Comment by dckc** [comment]
I still see new tests in `packages/ERTP/test/unitTests/mathHelpers/test-copySetMathHelpers.js` etc.

Odd... I don't see any existing tests for `AmountMath.isGTE` separate from the `MathHelpers` tests. It looks like we don't have feature tests for the external API. I need to check with @erights about this.

**Comment by dckc** [comment]
`test-natMathHelpers.js` and `test-setMathHelpers.js` are both at the MathHelpers level. Nothing in this PR should touch that layer of abstraction.

**Comment by b4d2** [comment]
Condensed the tests and move moved around the refactor of types.js - should be good to squash/rebase/merge

File: packages/ERTP/src/amountMath.js [hunk 2]
```diff
@@ -361,6 +361,26 @@ const AmountMath = {
...
+  /**
+   * Returns the min value between x and y using isGTE
+   *
+   * @template {AmountValue} [V=AmountValue]
+   * @param {Amount<V>} x
+   * @param {Amount<V>} y
+   * @param {Brand=} brand
+   * @returns {Amount<V>}
+   */
+  min: (x, y, b = undefined) => (AmountMath.isGTE(x, y, b) ? y : x),
```
Comment (dckc) [inline packages/ERTP/src/amountMath.js]:
```suggestion
  min: (x, y, brand = undefined) => (AmountMath.isGTE(x, y, brand) ? y : x),
```

File: packages/ERTP/src/amountMath.js [hunk 3]
```diff
@@ -361,6 +361,26 @@ const AmountMath = {
...
+  /**
+   * Returns the max value between x and y using isGTE
+   *
+   * @template {AmountValue} [V=AmountValue]
+   * @param {Amount<V>} x
+   * @param {Amount<V>} y
+   * @param {Brand=} brand
+   * @returns {Amount<V>}
+   */
+  max: (x, y, b = undefined) => (AmountMath.isGTE(x, y, b) ? x : y),
```
Comment (dckc) [inline packages/ERTP/src/amountMath.js]:
```suggestion
  max: (x, y, brand = undefined) => (AmountMath.isGTE(x, y, brand) ? x : y),
```

File: packages/ERTP/test/unitTests/test-amountProperties.js [hunk 4]
```diff
@@ -84,3 +84,12 @@ test('subtract: (x + y) - y = x; (y - x) + x = y if y >= x', t => {
     }),
   );
 });
+
+test('minmax', t => {
+  fc.assert(
+    fc.property(fc.record({ x: arbAmount, y: arbAmount }), ({ x, y }) => {
+      t.deepEqual(m.min(x, y), m.isGTE(x, y) ? y : x, 'SET: min');
+      t.deepEqual(m.max(x, y), m.isGTE(x, y) ? x : y, 'SET: max');
```
Comment (dckc) [inline packages/ERTP/test/unitTests/test-amountProperties.js:149]: It's a little wonky for the test code to be exactly the same as the implementation, but I suppose it doesn't hurt, and it does achieve the goal of testing that the API exists.

**Review by dckc — APPROVED** [review]
Looks good.

There are some lint warnings; let me see if I can take care of those with a few clicks...

**Pull request merged**